src/utils/PluginManager.ts

```diff
@@ -0,0 +1,123 @@
+import { PluginConfiguration } from '../fshtypes';
+import { logger } from './FSHLogger';
+
+export const AVAILABLE_HOOKS = [
+  'beforeFillTank',
+  'beforeLoadCustomResources',
+  'afterLoadCustomResources',
+  'afterExportFHIR'
+];
+
+export class PluginManager {
+  static hooks = new Map<string, ((...args: any) => any)[]>();
+
+  static async processPluginConfiguration(
+    plugins: PluginConfiguration[],
+    pluginBasePath: string
+  ): Promise<void> {
+    for (const plugin of plugins) {
+      // if the plugin name is given as name@version, it's something we may need to load from npm
+      // otherwise, we load it from the filesystem
+      // when we load from the filesystem, it might be at pluginBasePath, or at pluginBasePath/node_modules
+      const versionSeparator = plugin.name.lastIndexOf('@');
+      let pluginName: string;
+      let pluginVersion: string;
+      if (versionSeparator > 0) {
+        pluginName = plugin.name.slice(0, versionSeparator);
+        pluginVersion = plugin.name.slice(versionSeparator + 1);
+      } else {
+        pluginName = plugin.name;
+      }
+      try {
+        if (pluginVersion != null) {
+          // this is a plugin that can be acquired from npm.
+          // it might already be installed, so check for that first.
+          const existingInstallationPath = path.join(pluginBasePath, 'node_modules', pluginName);
+          if (fs.existsSync(path.join(existingInstallationPath, 'package.json'))) {
+            // check the version number in package.json
+            const packageContents = fs.readJsonSync(
+              path.join(existingInstallationPath, 'package.json')
+            );
+            if (packageContents.version === pluginVersion) {
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            } else {
+              await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+              PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+            }
+          } else {
+            await PluginManager.installFromNpm(pluginBasePath, plugin.name);
+            PluginManager.loadFromFilesystem([existingInstallationPath], pluginName, plugin.args);
+          }
+        } else {
+          // this is a plugin managed on the local filesystem.
+          // it may be at the base path, or in the node_modules directory
+          PluginManager.loadFromFilesystem(
+            [
+              path.join(pluginBasePath, pluginName),
+              path.join(pluginBasePath, 'node_modules', pluginName)
+            ],
+            pluginName,
+            plugin.args
+          );
+        }
+      } catch (err) {
+        logger.error(`Couldn't load plugin ${plugin.name}: ${err.message}`);
+      }
+    }
+  }
+
+  // a small test for name safety, based on typical naming patterns for npm packages
+  static isNameNpmSafe(name: string): boolean {
+    return /^[\w\d@/._-]+$/.test(name);
```

Comment:
Is this to avoid someone making a name like `foo@1.0.0 && sudo rm -rf /*`?

Reply:
That's correct: since this is user input, I don't want to hand it to a shell without checking it first.